boot/systems.go

```diff
@@ -20,9 +20,127 @@
+	}
+
+	status := vars["recovery_system_status"]
+	switch status {
+	case "":
+		// not trying any recovery systems right now
+		return false, nil
+	case "try", "tried":
+		// both are valid options, where tried may indicate there was an
+		// unexpected reboot somewhere along the path of getting back to
+		// the run system
+	default:
+		return false, &errInconsistentRecoverySystemState{
+			why: fmt.Sprintf("unexpected recovery system status %q", status),
+		}
+	}
+
+	trySystem := vars["try_recovery_system"]
+	if trySystem == "" {
+		// XXX: could we end up with one variable set and the other not?
+		return false, &errInconsistentRecoverySystemState{
+			why: fmt.Sprintf("try recovery system is unset but status is %q", status),
+		}
+	}
+
+	if trySystem == currentSystemLabel {
+		// we are running a recovery system indicated in the boot
+		// variables, which may or may not be considered good at this
+		// point, nonetheless we are in recover mode and thus consider
+		// the system as being tried
+
+		// note, with status set to ''tried', we may be back to the
+		// tried system again, most likely due to an unexpected reboot
+		// when coming back to run mode
+		return true, nil
+	}
+	// we may still be running an actual recovery system if such mode was
+	// requested
+	return false, nil
+}
+
+type TryRecoverySystemOutcome int
+
+const (
+	TryRecoverySystemOutcomeFailure TryRecoverySystemOutcome = iota
+	TryRecoverySystemOutcomeSuccess
+	// TryRecoverySystemOutcomeInconsistent indicates that the booted try
+	// recovery system state was incorrect and corresponding boot variables
+	// need to be cleared
+	TryRecoverySystemOutcomeInconsistent
+)
+
+// EnsureNextBootToRunModeWithTryRecoverySystemOutcome typically called while in
+// initramfs, updates the boot environment to indicate that the outcome of
+// trying out a recovery system and sets up the system to boot into run mode. It
+// is up to the caller to ensure the status is updated for the right recovery
+// system, typically by calling InitramfsIsTryingRecoverySystem beforehand.
+func EnsureNextBootToRunModeWithTryRecoverySystemOutcome(outcome TryRecoverySystemOutcome) error {
+	opts := &bootloader.Options{
+		// setup the recovery bootloader
+		Role: bootloader.RoleRecovery,
+	}
+	// TODO:UC20: seed may need to be switched to RW
+	bl, err := bootloader.Find(InitramfsUbuntuSeedDir, opts)
+	if err != nil {
+		return err
+	}
+	vars := map[string]string{
+		// always going to back to run mode
+		"snapd_recovery_mode":    "run",
+		"snapd_recovery_system":  "",
+		"recovery_system_status": "try",
+	}
+	switch outcome {
+	case TryRecoverySystemOutcomeFailure:
+		// already set up for this scenario
+	case TryRecoverySystemOutcomeSuccess:
+		vars["recovery_system_status"] = "tried"
+	case TryRecoverySystemOutcomeInconsistent:
+		// there may be an unexpected status, or the tried system label
+		// is unset, in either case, clear the status
+		vars["recovery_system_status"] = ""
```

Comment:
we could but are not forced to do it here, but we should remember to clear  try_recovery_system during run mode in this case